wrc/codegen/merger.py

```diff
@@ -0,0 +1,177 @@
+    """
+    Returns True if num1 < num2, otherwise returns False.
+    We need this for the same reasons we need compare_article_numbers().
+    """
+    for n1, n2 in zip(num1, num2):
+        if n1 < n2:
+            return True
+        elif n1 > n2:
+            return False
+
+    return len(num1) < len(num2)
+
+
+def get_reg_number(reg):
+    if isinstance(reg, Guideline):
+        reg = reg.number.split('+')[0]
+        num = split_rule_number(reg)
+    else:
+        num = split_rule_number(reg.number)
+    if str(num[-1]) == '0':
+        # Because split_rule_number() sometimes returns a 0 at the end, and we don't want that here.
+        del num[-1]
+    return num
+
+
+def merge_ast(astreg, astguide, language_options):
+    """
+    Combine the two AST into one.
+    """
+    # We only want the Articles, so we will start the for loop below at FIRST_ARTICLE_INDEX,
+    # and check if the indices for Notes, TOC and First Article are correct.
+    if not isinstance(astreg.sections[NOTES_INDEX], Section) or \
+            not isinstance(astguide.sections[NOTES_INDEX], Section):
+        raise BadFormatError('NOTES_INDEX is incorrect')
+    if not isinstance(astreg.sections[TOC_INDEX], TableOfContent):
+        raise BadFormatError('TOC_INDEX is incorrect')
+    if not (hasattr(astreg.sections[FIRST_ARTICLE_INDEX], 'number') and astreg.sections[
+                FIRST_ARTICLE_INDEX].number == '1'):
+        raise BadFormatError('FIRST_ARTICLE_INDEX is incorrect.')
+
+    # We take the astreg as the base, and then we add the guidelines in their corresponding places.
+    ast_combined = astreg
+
+    # Merge the notes:
+    # We won't include the "WCA Regulations" notes from the Guidelines, as it is the same as in
+    # "WCA Regulations and Guidelines" of the Regulations document.
+    # languages_options is used to handle translations.
+    ast_combined.sections[NOTES_INDEX].content.append(
+        [cont for cont in astguide.sections[NOTES_INDEX].content if language_options['regulations'] not in cont.title])
+
+    # Now we iterate over all the Articles found in astguide, starting from FIRST_ARTICLE_INDEX.
+    for guide_section in astguide.sections[FIRST_ARTICLE_INDEX:]:
+
+        stack = deepcopy(guide_section.content)
+        reg_section = None
+        # Find the associated section in the Regulations or add it if there isn't one.
+        for index, section in enumerate(ast_combined.sections[FIRST_ARTICLE_INDEX:]):
+            if not isinstance(section, Article):
+                # We could just continue here, but finding a non-Article section after FIRST_ARTICLE_INDEX
+                # is worth raising an exception.
+                raise BadFormatError('Section is not an Article.')
+
+            if section.number == guide_section.number:
+                # We found the section!
+                reg_section = section
+                break
+            elif compare_article_numbers(guide_section.number, section.number):
```

Comment:
I feel like the naming of the method is not optimal here. "If please compare article numbers" does not seem intuitive -- usually a comparison result is something like "a is smaller", "b is smaller" or "both are equal". How does this interact with an `if`?

Reply:
Looking at the source code of `compare_article_numbers` you might want to rename it to something like `article_numbers_ordered`, which reads more fluently: "If the article numbers are ordered (i.e. in the correct order), then..."

Reply:
Agreed, the name of the method is not clear at all. I renamed it to `article_number_is_lower`, which I think is the most intuitive option.

Related note:

Now I also understand how the fact that the parameters in `compare_reg_numbers` are lists might be confusing. Let me explain the purpose of that method (I changed its name to `reg_number_is_lower`):

The `split_rule_number` method used in our `get_reg_number` method, takes a string like `"12a5b"` and returns a list `[12, "a", 5, "b"]` (notice how it's made of integers and strings). We then use this list in the `reg_number_is_lower` method (formerly `compare_reg_numbers`).